### PR TITLE
[event_v2] indexer logic update to handle migration (#360)

### PR DESCRIPTION
**Pull request by lightmark**
revert https://github.com/aptos-labs/aptos-indexer-processors/commit/8c41d26171446837edcbf6c7a4084e59fd43f4ba
with a fix

### CODE REVIEW

File: rust/processor/src/models/coin_models/coin_utils.rs [hunk 1]
```diff
@@ -288,16 +302,38 @@ impl CoinResource {
...
+        txn_version: i64,
+    ) -> Result<Option<(CoinEvent, Option<String>)>> {
         match data_type {
-            "0x1::coin::WithdrawEvent" => {
-                serde_json::from_str(data).map(|inner| Some(CoinEvent::WithdrawCoinEvent(inner)))
+            "0x1::coin::WithdrawEvent" => serde_json::from_str(data)
+                .map(|inner| Some((CoinEvent::WithdrawCoinEvent(inner), None))),
+            "0x1::coin::DepositEvent" => serde_json::from_str(data)
+                .map(|inner| Some((CoinEvent::WithdrawCoinEvent(inner), None))),
+            t if t.starts_with("0x1::coin::Withdraw") => {
```
Comment (rtso) [inline rust/processor/src/models/coin_models/coin_utils.rs:320]: why not explicitly match the event name?

File: rust/processor/src/models/mod.rs [hunk 2]
```diff
@@ -16,3 +20,111 @@ pub mod token_models;
...
+            "0x3::token_transfers::TokenClaimEvent",
+            "0x3::token_transfers::TokenClaim",
+        );
+        m.insert("0x4::token::MutationEvent", "0x4::token::Mutation");
+        m.insert("0x1::object::TransferEvent", "0x1::object::TransferEvent");
+        m
+    };
+    pub static ref EVENTS_MIGRATION_PREFIX_MAP: HashMap<&'static str, &'static str> = {
+        let mut m = HashMap::new();
+        m.insert("0x1::coin::WithdrawEvent", "0x1::coin::Withdraw<");
```
Comment (rtso) [inline rust/processor/src/models/mod.rs:77]: Can you add a comment with example of what this is for?

File: rust/processor/src/models/mod.rs [hunk 3]
```diff
@@ -16,3 +20,111 @@ pub mod token_models;
...
+        m.insert("0x1::coin::DepositEvent", "0x1::coin::Deposit<");
+        m
+    };
+}
+
+pub(crate) fn should_skip(index: usize, event: &Event, events: &[Event]) -> bool {
+    index > 0
+        && (EVENTS_MIGRATION_MAP.get(event.type_str.as_str())
+            == Some(&&*events[index - 1].type_str)
```
Comment (rtso) [inline rust/processor/src/models/mod.rs:86]: Can you add a comment with the assumption this is making? I'm guessing the assumption is that an event v2 will always come right after an event v1?

File: rust/processor/src/models/mod.rs [hunk 4]
```diff
@@ -16,3 +20,111 @@ pub mod token_models;
...
+    };
+    pub static ref EVENTS_MIGRATION_PREFIX_MAP: HashMap<&'static str, &'static str> = {
+        let mut m = HashMap::new();
+        m.insert("0x1::coin::WithdrawEvent", "0x1::coin::Withdraw<");
+        m.insert("0x1::coin::DepositEvent", "0x1::coin::Deposit<");
+        m
+    };
+}
+
+pub(crate) fn should_skip(index: usize, event: &Event, events: &[Event]) -> bool {
```
Comment (rtso) [inline rust/processor/src/models/mod.rs:83]: The logic may be simpler if we do a first pass of the events to check if it's events v1 or v2. Then should_skip(is_events_v2, event_name) just checks if the event_name is in the v1 set or v2 set.

File: rust/processor/src/processors/fungible_asset_processor.rs [hunk 5]
```diff
@@ -433,6 +434,9 @@ async fn parse_v2_coin(
 
         // Loop to handle events and collect additional metadata from events for v2
         for (index, event) in events.iter().enumerate() {
+            if should_skip(index, event, events) {
+                continue;
+            };
```
Comment (rtso) [inline rust/processor/src/processors/fungible_asset_processor.rs:439]: confirmed with @lightmark we don't need this for fungible asset

File: rust/processor/src/models/mod.rs (hunk 2 above)
Comment (rtso) [inline rust/processor/src/models/mod.rs:77]: What's the rationale for having coin type generic in the event name? This makes it so much harder to index bc we need to parse out the generic part of the event name. Why not have coin_type as part of the event struct?

**Review by bowenyang007 — CHANGES_REQUESTED** [review]
pushing back to your queue. Is it still relevant? If so let's address the comments.

File: rust/processor/src/processors/token_v2_processor.rs [hunk 6]
```diff
@@ -770,41 +771,66 @@ async fn parse_v2_token(
...
+                            // note that these will be multiplied by -1 so that it doesn't conflict with wsc index
+                            let index = if index == 0 {
+                                user_txn.events.len()
+                            } else {
+                                index
+                            };
+                            aggregated_data
+                                .transfer_events
+                                .push((index as i64, transfer_events));
+                        }
```
Comment (rtso) [inline rust/processor/src/processors/token_v2_processor.rs:831]: Why is this changed so much? I thought `should_skip` handles skipping the v1 events

**Review by rtso — CHANGES_REQUESTED** [review]

File: rust/processor/src/db/common/models/mod.rs [hunk 7]
```diff
@@ -16,3 +23,63 @@ pub mod token_models;
 pub mod token_v2_models;
 pub mod transaction_metadata_model;
 pub mod user_transactions_models;
+
+static EVENT_V2_ENABLED: AtomicBool = AtomicBool::new(false);
```
[the remark this answers is not in the document]
Comment (lightmark) [reply rust/processor/src/db/common/models/mod.rs:27]: how do you know whether to skip it when you see a v1?

File: rust/processor/src/processors/fungible_asset_processor.rs [hunk 8]
```diff
@@ -596,6 +597,9 @@ async fn parse_v2_coin(
 
         // Loop to handle events and collect additional metadata from events for v2
         for (index, event) in events.iter().enumerate() {
+            if should_skip(event) {
```
[the remark this answers is not in the document]
Comment (lightmark) [reply rust/processor/src/processors/fungible_asset_processor.rs:600]: yes. so it does not hurt. This line is added by Larry :(

File: rust/processor/src/processors/token_v2_processor.rs (hunk 6 above)
Comment (lightmark) [reply rust/processor/src/processors/token_v2_processor.rs:831]: ...Because... Igor makes them mixed now. go to the smart contracts and then you will know why. That's why I want to separate a fix for kent first.

File: rust/processor/src/db/common/models/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (rtso) [reply rust/processor/src/db/common/models/mod.rs:27]:
```
let mapping = {
 withdrawV1 -> withdrawV2
}
should_skip(event_name, prev_event_name):
if event_name is in mapping and mapping[event_name] == prev_event_name: true
else: false

```

File: rust/processor/src/processors/token_v2_processor.rs (hunk 6 above)
Comment (rtso) [reply rust/processor/src/processors/token_v2_processor.rs:831]: Chatted offline about this. There could be v1 and v2 event next to each other for different collections and could accidentally skip a v1 event 

To get around this, we'll create `should_skip` that only skips if event names AND collection address matches.

File: python/processors/nft_orderbooks/parsers/okx_parser.py
Comment (bowenyang007) [inline python/processors/nft_orderbooks/parsers/okx_parser.py:1]: can we remove these? we're no longer maintaining the python stuff anyway so it's not needed and we don't want to test it anyway.

Comment (lightmark) [reply python/processors/nft_orderbooks/parsers/okx_parser.py:1]: larry added these. 
what to do with this file?

Comment (bowenyang007) [reply python/processors/nft_orderbooks/parsers/okx_parser.py:1]: uh just revert to how it was.

**Comment by bowenyang007** [comment]
Need a better summary and test plan. The summary cannot be "revert of a revert".

File: rust/processor/src/db/common/models/coin_models/coin_activities.rs [hunk 9]
```diff
@@ -186,6 +187,16 @@ impl CoinActivity {
...
+            if should_skip(
+                event,
+                if index == 0 {
+                    None
+                } else {
+                    Some(&events[index - 1])
+                },
+            ) {
+                continue;
+            }
```
Comment (bowenyang007) [inline rust/processor/src/db/common/models/coin_models/coin_activities.rs:199]: What's happening here?

File: rust/processor/src/db/common/models/coin_models/coin_activities.rs [hunk 10]
```diff
@@ -222,23 +233,44 @@ impl CoinActivity {
...
+                inner.amount.clone(),
+                Some(inner.coin_type.clone()),
+            ),
         };
-        let coin_type =
+        let coin_type = if let Some(coin_type) = coin_type_option {
+            coin_type
+        } else {
+            let event_move_guid = EventGuidResource {
+                addr: standardize_address(event.key.as_ref().unwrap().account_address.as_str()),
```
Comment (bowenyang007) [inline rust/processor/src/db/common/models/coin_models/coin_activities.rs:262]: this is just owner_account from above right?

File: rust/processor/src/db/common/models/coin_models/coin_activities.rs [hunk 11]
```diff
@@ -222,23 +233,44 @@ impl CoinActivity {
...
+                Some(inner.coin_type.clone()),
+            ),
+            CoinEvent::DepositCoinEventV2(inner) => (
+                standardize_address(&inner.account),
+                inner.amount.clone(),
+                Some(inner.coin_type.clone()),
+            ),
         };
-        let coin_type =
+        let coin_type = if let Some(coin_type) = coin_type_option {
```
Comment (bowenyang007) [inline rust/processor/src/db/common/models/coin_models/coin_activities.rs:258]: So for v2 what happens?

**Comment by lightmark** [comment]
we will not do double emission.
refactored.

**Review by rtso — APPROVED** [review]
